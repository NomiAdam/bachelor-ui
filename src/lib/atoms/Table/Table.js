import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export const StyledTable = styled.table`
	width: 100%;
	border-collapse: ${({rowSpacing}) => rowSpacing ? 'separate' : 'collapse'};
	padding-right: ${({rowSpacing}) => rowSpacing ? '5px' : '0'};
	border-spacing: ${({rowSpacing}) => rowSpacing ? '0 5px' : '0'};
	font-size: .85em;
`;

class Table extends PureComponent {
    render() {
        const {children, rowSpacing} = this.props;
        return (
            <StyledTable rowSpacing={rowSpacing}>
                {children}
            </StyledTable>
        )
    }
}

Table.propTypes = {
    /**
     * Children nodes of Table, Table body is expected
     */
    children: PropTypes.any,
    /**
     * Whether we should have space between rows or not
     */
    rowSpacing: PropTypes.bool,
};

Table.defaultProps = {
    rowSpacing: false,
};

export default Table;
