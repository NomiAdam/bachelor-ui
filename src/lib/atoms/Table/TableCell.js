import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledTableCell = styled.td`
	max-width: 100%;
	padding: 10px;
	${({editable}) => editable ? 'max-width: 100px' : ''};
	&:hover {
	  cursor: ${({redirect}) => redirect ? 'pointer' : 'inherit' };
	}
`;

class TableCell extends PureComponent {
	render() {
		const { children, colSpan, redirect, onClick, editable } = this.props;
		return(
			<StyledTableCell
				colSpan={colSpan}
				redirect={redirect}
                onClick={onClick}
                editable={editable}
			>
				{ children }
			</StyledTableCell>
		)
	}
}

TableCell.propTypes = {
	/**
	 * Number representation space between Cells
	 */
	colSpan:PropTypes.number,
	/**
	 * Children nodes of TableCell component
	 */
	children: PropTypes.any,
    /**
     * Whether we redirect on click or not
     */
    redirect: PropTypes.bool,
    /**
	 * Function to be called upon TableCell Click
     */
    onClick: PropTypes.func,
    /**
	 * Whether TableCell is used for Edit Icons
     */
    editable: PropTypes.bool,
};

TableCell.defaultProps = {
    redirect: false,
};

export { StyledTableCell };
export default TableCell;
