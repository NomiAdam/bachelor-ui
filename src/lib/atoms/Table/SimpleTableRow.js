import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { map, addIndex } from 'ramda';
import TableCell from './TableCell';

const mapIndexed = addIndex(map);

const StyledTableRow = styled.tr`
	background-color: transparent;
	td:first-child {
	  float: left;
	}
`;

const mapTableCells = (value, index) => <TableCell key={ index }>{value}</TableCell>;
const mappedTableCells = mapIndexed(mapTableCells);
class SimpleTableRow extends PureComponent {
    state = {
        rowData: [],
    };

    componentDidMount() {
        this.handleTableCellsMap();
    }

    componentDidUpdate(prevProps) {
        const { simpleTableRowData } = this.props;
        if (JSON.stringify(simpleTableRowData) !== JSON.stringify(prevProps.simpleTableRowData)) {
            this.handleTableCellsMap();
        }
    }

    handleTableCellsMap() {
        const { simpleTableRowData } = this.props;
        this.setState({ rowData: mappedTableCells(simpleTableRowData) });
    }

    render() {
        const { labeled, label } = this.props;
        const { rowData } = this.state;
        return (
            <StyledTableRow>
                {labeled && <TableCell>{label}</TableCell>}
                {rowData}
            </StyledTableRow>
        );
    }
}

SimpleTableRow.propTypes = {
    /**
     * Array of values is expected, value can be string, number or another Component
     */
    simpleTableRowData: PropTypes.array,
    /**
     * Boolean value, when true extra TabelCell is added to the row
     */
    labeled: PropTypes.bool,
    /**
     * Can be any value, be it string, number or Component, is displayed only when Row is labeled
     */
    label: PropTypes.any,
};

SimpleTableRow.defaultProps = {
    simpleTableRowData: [],
    labeled: false,
};

export { StyledTableRow };
export default SimpleTableRow;
