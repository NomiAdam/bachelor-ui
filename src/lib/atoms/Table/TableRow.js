import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { compose, tail, values } from 'ramda';
import { mapIndexed } from 'ramda-extension';
import { IoEdit, IoAndroidDelete } from 'react-icons/lib/io';
import TableCell from './TableCell';
import { lightTheme, darkTheme } from '../../constants/theme';

const StyledTableRow = styled.tr`
	background-color: transparent;
	border-bottom: 1px solid ${ lightTheme.lightGrey };
	transition: .25s all ease;
	&:hover {
		background-color: ${ ({ effect }) => (effect ? 'transparent' : lightTheme.whiteBlue) };
	}
`;

const Icon = styled.div`
    padding: 10px 15px;
    font-size: 1.5em;
    cursor: pointer; 
    color: ${ darkTheme.darkGrey };
`;

const StyledEdit = Icon.withComponent(IoEdit);
const StyledTrash = Icon.withComponent(IoAndroidDelete);

class TableRow extends PureComponent {
    state = {
        rowData: [],
    };

    componentDidMount() {
        this.handleTableCellsMap();
    }

    componentDidUpdate(prevProps) {
        const { tableRowData } = this.props;
        if (tableRowData !== prevProps.tableRowData) {
            this.handleTableCellsMap();
        }
    }

    handleTableCellsMap = () => {
        const { tableRowData, redirect, handleRedirect } = this.props;

        const mapTableCells = (value, index) => (
            <TableCell
                redirect={ redirect }
                key={ `${ index }-${ value }` }
                onClick={ handleRedirect }
            >
                {value}
            </TableCell>
        );
        const mappedTableCells = compose(tail, values, mapIndexed(mapTableCells));

        this.setState({ rowData: mappedTableCells(tableRowData) });
    };

    render() {
        const { editable, handleDelete, handleEdit } = this.props;
        const { rowData } = this.state;
        return (
            <StyledTableRow>
                {editable && (
                    <TableCell editable>
                        <StyledTrash onClick={ handleDelete } />
                        <StyledEdit onClick={ handleEdit } />
                    </TableCell>
                )}
                {rowData}
            </StyledTableRow>
        );
    }
}

TableRow.propTypes = {
    /**
     * Whether Table row should have something of Effect on hover, in our case color change
     */
    effect: PropTypes.bool,
    /**
     * Object of keys is expected
     */
    tableRowData: PropTypes.object,
    /**
     *  Whether table has extra column with edit Icons, boolean is expected
     */
    editable: PropTypes.bool,
    /**
     * Function to handle edit icon action, is present only when table is editable
     */
    handleEdit: PropTypes.func,
    /**
     * Function to handle delete icon action, is present only when table is editable
     */
    handleDelete: PropTypes.func,
    /**
     * Whether we redirect on click or not
     */
    redirect: PropTypes.bool,
    /**
     * Function to handle redirect on click, if redirect is set to false, handleRedirect is null
     */
    handleRedirect: PropTypes.any,
};

TableRow.defaultProps = {
    tableRowData: {},
    redirect: false,
    editable: false,
};

export { StyledTableRow, StyledEdit, StyledTrash };
export default TableRow;
