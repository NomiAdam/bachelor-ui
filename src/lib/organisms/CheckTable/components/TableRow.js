import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {TableCell} from '../../../atoms/Table/index';
import {StyledTableRow} from '../../../atoms/Table/SimpleTableRow';
import Checkbox from '../../../atoms/Checkbox/index';
import CheckCell from './CheckCell';
import {
    map, addIndex, indexOf, o, append, equals, ifElse, __, remove, always, reduce, add,
} from 'ramda';

const mapIndexed = addIndex(map);
const isSelected = equals(-1);

const computeSelectedArray = reduce(add, 0);

class TableRow extends PureComponent {
    state = {
        rowData: [],
    };

    componentDidMount() {
        this.handleDayHoursMap();
    };

    componentDidUpdate(prevProps) {
        const {selected} = this.props;
        if (computeSelectedArray(selected) !== computeSelectedArray(prevProps.selected)) {
            this.handleDayHoursMap();
        }
    };

    handleClick = (id) => (e) => {
        e.preventDefault();

        const {selected, handleChange} = this.props;

        const indexOfSelected = indexOf(__, selected);

        const appendWithSelected = always(append(id, selected));
        const removeFromSelected = always(o(remove(__, 1, selected), indexOfSelected)(id));

        const removeOrConcat = ifElse(isSelected, appendWithSelected, removeFromSelected);

        const changedSelectedState = o(removeOrConcat, indexOfSelected);

        handleChange(changedSelectedState(id));
    };

    handleDayHoursMap = () => {
        const {dayHours, isCheckbox} = this.props;
        const mapDayHour = (hour, index) => {
            return isCheckbox ?
                (
                    <TableCell key={index} onClick={this.handleClick(index)}>
                        <Checkbox checked={this.isSelected(index) && !this.isDisabled(index)}/>
                    </TableCell>
                ) : (
                    <CheckCell checked={this.isSelected(index)} disabled={this.isDisabled(index)}/>
                );
        };
        const mappedReservationCells = mapIndexed(mapDayHour);

        this.setState({rowData: mappedReservationCells(dayHours)});
    };

    isSelected = id => this.props.selected.indexOf(id) !== -1;

    isDisabled = id => this.props.disabled.indexOf(id) !== -1;

    render() {
        const {label} = this.props;
        const {rowData} = this.state;
        return (
            <StyledTableRow>
                <TableCell>
                    {label}
                </TableCell>
                {rowData}
            </StyledTableRow>
        );
    }
}

TableRow.propTypes = {
    /**
     * String representing label for each row, note that this can be day or sports center
     */
    label: PropTypes.string.isRequired,
    /**
     * Array of displayed day hours
     */
    dayHours: PropTypes.array,
    /**
     * Handle onClick on TableCell, (value) => is expected
     */
    handleChange: PropTypes.func,
    /**
     * Array of checked TableCells for current row
     */
    selected: PropTypes.array,
    /**
     * Whether we use checkbox inside component or not
     */
    isCheckbox: PropTypes.bool,
    /**
     * Array of disabled TableCells for current row
     */
    disabled: PropTypes.array,
};

TableRow.defaultProps = {
    dayHours: [],
    selected: [],
    disabled: [],
    isCheckbox: true,
};

export default TableRow;
