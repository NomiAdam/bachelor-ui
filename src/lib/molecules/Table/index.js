import TableCell from './TableCell';
import TableRow from './TableRow';
import SimpleTableRow from './SimpleTableRow';
import Table from './Table';

export {
    TableCell,
    TableRow,
    SimpleTableRow,
};

export default Table;
