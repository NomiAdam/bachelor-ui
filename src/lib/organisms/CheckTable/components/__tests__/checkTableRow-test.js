import React from 'react';
import { shallow } from 'enzyme';
import TableRow from '../TableRow';
import {TableCell} from '../../../../atoms/Table/index';
import {StyledTableRow} from '../../../../atoms/Table/SimpleTableRow';

const mockSelected = [0, 1];

const mockDayHours = [
    '06:00',
    '06:30',
    '07:00',
    '07:30',
    '08:00',
    '08:30',
];

const mockLabel = 'label';

describe('CheckTableRow', () => {
    it('should be defined', () => {
        expect(shallow(<TableRow label={mockLabel}/>)).toBeDefined();
    });
    it('should match the snapshot', () => {
        expect(shallow(<TableRow label={mockLabel}/>)).toMatchSnapshot();
    });
    it('should match the snapshot with values', () => {
        expect(shallow(<TableRow dayHours={mockDayHours} label={mockLabel}
                                 selected={mockSelected}/>)).toMatchSnapshot();
    });
    it('should be of type StyledTableRow', () => {
        expect(shallow(<TableRow label={mockLabel}/>).type()).toBe(StyledTableRow);
    });
    it('should have render when props are not the same TableCell with length of mockDayHours + 1 as label', () => {
        const wrapper = shallow(<TableRow label={mockLabel} dayHours={mockDayHours}/>);
        wrapper.setProps({ selected: mockSelected });
        expect(wrapper.find(TableCell).length).toBe(mockDayHours.length + 1);
    });
});
