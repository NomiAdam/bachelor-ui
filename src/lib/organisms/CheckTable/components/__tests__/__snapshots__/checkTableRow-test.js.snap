// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckTableRow should match the snapshot 1`] = `
<SimpleTableRow__StyledTableRow>
  <TableCell
    redirect={false}
  >
    label
  </TableCell>
</SimpleTableRow__StyledTableRow>
`;

exports[`CheckTableRow should match the snapshot with values 1`] = `
<SimpleTableRow__StyledTableRow>
  <TableCell
    redirect={false}
  >
    label
  </TableCell>
  <TableCell
    key="0"
    onClick={[Function]}
    redirect={false}
  >
    <Checkbox
      checked={true}
      labelPosition="LEFT"
    />
  </TableCell>
  <TableCell
    key="1"
    onClick={[Function]}
    redirect={false}
  >
    <Checkbox
      checked={true}
      labelPosition="LEFT"
    />
  </TableCell>
  <TableCell
    key="2"
    onClick={[Function]}
    redirect={false}
  >
    <Checkbox
      checked={false}
      labelPosition="LEFT"
    />
  </TableCell>
  <TableCell
    key="3"
    onClick={[Function]}
    redirect={false}
  >
    <Checkbox
      checked={false}
      labelPosition="LEFT"
    />
  </TableCell>
  <TableCell
    key="4"
    onClick={[Function]}
    redirect={false}
  >
    <Checkbox
      checked={false}
      labelPosition="LEFT"
    />
  </TableCell>
  <TableCell
    key="5"
    onClick={[Function]}
    redirect={false}
  >
    <Checkbox
      checked={false}
      labelPosition="LEFT"
    />
  </TableCell>
</SimpleTableRow__StyledTableRow>
`;
